cdc/redo/reader/reader.go

```diff
@@ -145,203 +141,140 @@ func (l *LogReader) ResetReader(ctx context.Context, startTs, endTs uint64) erro
 		}
 	}
 
-	if startTs > endTs || startTs > l.meta.ResolvedTs || endTs <= l.meta.CheckpointTs {
-		return errors.Errorf(
-			"startTs, endTs (%d, %d] should match the boundary: (%d, %d]",
-			startTs, endTs, l.meta.CheckpointTs, l.meta.ResolvedTs)
-	}
-	return l.setUpReader(ctx, startTs, endTs)
+	eg, egCtx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		return l.runRowReader(egCtx)
+	})
+	eg.Go(func() error {
+		return l.runDDLReader(egCtx)
+	})
+	return eg.Wait()
 }
 
-func (l *LogReader) setUpReader(ctx context.Context, startTs, endTs uint64) error {
-	l.Lock()
-	defer l.Unlock()
-
-	var errs error
-	errs = multierr.Append(errs, l.setUpRowReader(ctx, startTs, endTs))
-	errs = multierr.Append(errs, l.setUpDDLReader(ctx, startTs, endTs))
-
-	return errs
-}
-
-func (l *LogReader) setUpRowReader(ctx context.Context, startTs, endTs uint64) error {
-	l.rowLock.Lock()
-	defer l.rowLock.Unlock()
-
-	err := l.closeRowReader()
-	if err != nil {
-		return err
-	}
-
+func (l *LogReader) runRowReader(egCtx context.Context) error {
+	defer close(l.rowCh)
 	rowCfg := &readerConfig{
-		startTs:            startTs,
-		endTs:              endTs,
+		startTs:            l.meta.CheckpointTs,
+		endTs:              l.meta.ResolvedTs,
 		dir:                l.cfg.Dir,
 		fileType:           redo.RedoRowLogFileType,
 		uri:                l.cfg.URI,
 		useExternalStorage: l.cfg.UseExternalStorage,
 		workerNums:         l.cfg.WorkerNums,
 	}
-	l.rowReader, err = newReader(ctx, rowCfg)
-	if err != nil {
-		return err
-	}
-
-	l.rowHeap = logHeap{}
-	l.cfg.startTs = startTs
-	l.cfg.endTs = endTs
-	return nil
+	return l.runReader(egCtx, rowCfg)
 }
 
-func (l *LogReader) setUpDDLReader(ctx context.Context, startTs, endTs uint64) error {
-	l.ddlLock.Lock()
-	defer l.ddlLock.Unlock()
-
-	err := l.closeDDLReader()
-	if err != nil {
-		return err
-	}
-
+func (l *LogReader) runDDLReader(egCtx context.Context) error {
+	defer close(l.ddlCh)
 	ddlCfg := &readerConfig{
-		startTs:            startTs,
-		endTs:              endTs,
+		startTs:            l.meta.CheckpointTs - 1,
```

Comment:
why is -1 needed?

Reply:
If there is a ddl event in the check point, there is no way to determine whether the event was executed. Therefore, it is a safe practice to always assume that the event did not execute successfully and try to execute this DDL event in applier, which is similar to initializing DDLPuller with `checkpointTs - 1` when owner restarts.